### ISSUE TITLE
Async python support: aiopynamodb 

### ISSUE THREAD

**Issue by kamadorueda** [pynamodb/PynamoDB#802]
https://github.com/aio-libs/aiobotocore
https://github.com/terrycain/aioboto3

**Comment by garrettheel** [comment]
The discussion here is relevant: https://github.com/pynamodb/PynamoDB/issues/525#issuecomment-607892448

I'd like to support asyncio natively in the library, but I'm still a little hesitant to adopt aiobotocore right as it's not maintained by AWS. We don't rely on all that much of botocore right now, so one option would be to drop that altogether and provide a separate async interface

**Comment by dwatkinsweb** [comment]
Any idea when this might happen? We could really use this feature right now. I've been attempting to do this myself but I've been having to duplicate a lot of your code for a few small changes.

**Comment by kamadorueda** [comment]
There is another approach that is used by many libraries out there (keep reading for examples):

When a library exposes a high-latency function, for instance:

```py
for item in TestModel.view_index.query(1):
    print("Item queried from index: {0}".format(item))
```

One can wrap the calls in a sub-thread via [loop.run_in_executor](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor).

Since that's is a little verbose there are nice libraries to make it human-friendly, for example [aioextensions](https://kamadorueda.github.io/aioextensions)

So the syntax would be something like:

```py
from aioextensions import in_thread

for item in await in_thread(TestModel.view_index.query, 1):
    print("Item queried from index: {0}".format(item))
```

Which would run the high-latency thing in a sub-thread that allows for concurrency.

It's a very minimalistic interface and requires no work from pynamodb since it's on the consumer side to do the wrapping:

```py
from aioextensions import in_thread, collect

# Equivalent to pynamodb_func(arg_1, arg_2, kwarg_a=3, kwarg_b=4)
one_query = await in_thread(pynamodb_func, arg_1, arg_2, kwarg_a=3, kwarg_b=4)

# Equivalent to pynamodb_func(arg_1, arg_2, kwarg_a=3, kwarg_b=4) but all queries concurrently (overlapping in time) and fast!!
many_queries = await collect([
    in_thread(pynamodb_func, arg_1, arg_2, kwarg_a=kwarg_a, kwarg_b=kwarg_b)
    for arg_1, arg_2, kwarg_a, kwarg_b in [long list of things to fetch]
])
```

There is another alternative and is providing _async versions of the functions, which internally could use the mentioned wrappers hiding them from the final user:

```py
def pynamodb_func(arg_1, arg_2, kwarg_a=3, kwarg_b=4) -> Data:
    ....

async def async_pynamodb_func(arg_1, arg_2, kwarg_a=3, kwarg_b=4) -> Data:
   return await in_thread(pynamodb_func, arg_1, arg_2, kwarg_a=kwarg_a, kwarg_b=kwarg_b)
```

The library also offers some nice helpers that we could find useful like workers, batching and rate limits.

I think I'm volunteering to implement the async wrappers if you think it's a nice approach, you tell me! @garrettheel 

These are examples of the mentioned sub-thread wrapping:

- [aiofiles](https://github.com/Tinche/aiofiles/blob/master/aiofiles/threadpool/__init__.py#L80)
- [integrates](https://gitlab.com/fluidattacks/product/-/blob/master/integrates/django-apps/integrates-back-async/backend/api/resolvers/vulnerability.py#L652)
- [python official docs example](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)

I've personally used it in production and the benefits from concurrency are worth the small overhead it adds to every call

It's common to use `a_`, `async_` or `_async` notation when both flavors are offered by a library

**Comment by garrettheel** [comment]
`loop.run_in_executor` is an interesting approach, but I have tried this before and seen performance issues with high-throughput applications using this approach. Introducing threads also introduces new and interesting failure modes that didn't exist before. I'd be concerned about going down that path, especially since the vast majority of users would still use the sync interface and pay that tax

I've been experimenting with a different approach in https://github.com/pynamodb/PynamoDB/pull/853, which could be characterized as a hackier version of the above suggestion (to the benefit of not requiring threads).

**Comment by brunobelloni** [comment]
Can also be done using `asyncio`. Will already be prepared for an eventual real `async` PynamoDB
Working on `Python 3.9.14+`

`asyncio.to_thread` uses `ThreadPoolExecutor` under the hood

```
import asyncio


async def main():
    # Equivalent to pynamodb_func(arg_1, arg_2, kwarg_a=3, kwarg_b=4)
    one_query = await asyncio.to_thread(pynamodb_func, arg_1, arg_2, kwarg_a=3, kwarg_b=4)

    # Equivalent to pynamodb_func(arg_1, arg_2, kwarg_a=3, kwarg_b=4) but all queries concurrently (overlapping in time) and fast!!
    many_queries = await asyncio.gather([
        asyncio.to_thread(pynamodb_func, arg_1, arg_2, kwarg_a=kwarg_a, kwarg_b=kwarg_b)
        for arg_1, arg_2, kwarg_a, kwarg_b in [long list of things to fetch]
    ])


if __name__ == '__main__':
    asyncio.run(main())
```

**Comment by aaronclong** [comment]
[@tasn](https://github.com/tasn) I notice you tried to do this with threading: https://github.com/pynamodb/PynamoDB/pull/968

**Comment by AbendGithub** [comment]
I am working on a project that we will benefit from adding async support to this package. We will implement our solution basically wrapping everything you have using Gevent. Why Gevent? Because you do not need to worry about async/await syntax, you do not need to rewrite everything defining async methods.

We will probably implement this before June, so as soon as I get some results from it, I will come back with a PR implementing it.

**Comment by aaronclong** [comment]
@AbendGithub I think long-term async/await is the future of python, though. Gevent isn't native or widely used by most python programmers.

**Comment by ikonst** [comment]
We use pynamodb with gevent pretty much everywhere at Lyft without any modifications to this library (with gevent standard monkey-patching).

There's been a lot of interest in adding an asyncio layer to this library over the years. It's not entirely trivial and will probably result in lots of duplication (seen this in redis-py).

I'd also see it as a negative testimony to the asyncio approach (aka blue/green functions), but this train left the station and we're mostly locked into the architectures we invested into, so I can definitely see the value in an asyncio layer.

**Comment by aaronclong** [comment]
Yeah, I know the blue/green function debate is quite polarizing. However, as you said, the language is natively adopting the once approach. Eventually, I feel like even boto3 will be forced to adopt asyncio.

**Comment by dbfreem** [comment]
Hey just curious if this ever caught traction. I feel like asyncio is one of the easiest ways I find to improve io bound apps.

**Comment by brunobelloni** [comment]
Hey guys!

Made a working aiobotocore fork of PynamoDB. Works on Python 3.10+ and aiobotocore>=2.19.0.

Basic stuff is tested and working (queries, scans, gets, puts etc). Need help testing and improving it!

Here's the repo: https://github.com/brunobelloni/AioPynamoDB

You can try it with:
```bash
pip install git+https://github.com/brunobelloni/AioPynamoDB#egg=aiopynamodb
```

**Comment by heitorlessa** [comment]
I've long admired PynamoDB but tbh they could take some stocks and inspiration from the likes of https://github.com/ferrumio/pydynox -- with Pyo3, you could have and eat the cake too (async+Rust perf)
